pkg/agent/client/client.go

```diff
@@ -616,29 +627,69 @@ func (c *client) fetchBundles(ctx context.Context, federatedBundles []string) ([
 	}
 	bundles = append(bundles, bundle)
 
-	for _, b := range federatedBundles {
-		federatedTD, err := spiffeid.TrustDomainFromString(b)
-		if err != nil {
-			return nil, err
+	return c.fetchBundlesConcurrently(ctx, bundleClient, federatedBundles, bundles)
+}
+
+// fetchBundlesConcurrently fetches federated bundles concurrently.
+// This is done to improve sync times when there are many federations. This should ensure that the
+// sync does not exceed rpcTimeout.
+func (c *client) fetchBundlesConcurrently(ctx context.Context, bundleClient bundlev1.BundleClient, trustDomains []string, bundles []*types.Bundle) ([]*types.Bundle, error) {
+	jobCh := make(chan string)
+	resultCh := make(chan fetchBundleResult, len(trustDomains))
+	// Start a set of worker goroutines.
+	wg := sync.WaitGroup{}
+	for range min(maxBundleWorkers, len(trustDomains)) {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
```

Comment:
I think you can now do something like this to avoid the Add() and defer:
```suggestion
		wg.Go(func() {
```